### PR TITLE
fix(tokenless): surface schema compression skips on tool-less BeforeModel

### PR DESCRIPTION
**Pull request by Forrest-ly**
## 改动说明

让 Schema 压缩 BeforeModel Hook 的跳过路径可诊断，并补充 Gemini 包装形态的增量测试与文档。

**范围说明**：Gemini `functionDeclarations` 包装的压缩器核心已由 #2663（`625f0779`）合入 main，本 PR 不包含该部分，仅保留 main 未覆盖的增量测试与文档（详见文末 rebase 记录）。

### 背景

统计只记录有实际 Token 节省的调用；而「事件不携带工具声明 / 负载不可解析」的跳过路径此前完全静默，「0 条 compress-schema 记录」无法与「Hook 未触发」区分，现场曾误判「cosh 没有发射 BeforeModel」。copilot-shell 的 Gemini 工具形态零收益问题（核心缺陷）已由 main 的 #2663 修复；本 PR 针对的是跳过路径的可诊断化。

### 变更内容

**A. 静默跳过可诊断化（Python hook）**

- BeforeModel payload 不是 JSON 对象、缺少 `llm_request`，或在 canonical / legacy 位置均找不到工具声明时，给出每会话一次的警告（stderr + Hook 输出 `systemMessage`，失败仍放行）；合法的空 canonical tools 数组保持静默；
- marker 记录有界的已告警 session key 集合（最新在最后，超限裁最旧），其读改写整体由专用 sidecar 锁文件上的独占 flock 串行化——同一 HOME 下并发会话的首次告警不再互相覆盖 key（修复前强制交错可 20/20 复现丢 key）；无 fcntl 平台或锁文件创建失败时退回无锁 best-effort 语义；
- 解析逻辑抽取为 `_extract_tools()`，canonical 优先、legacy 回退的既有行为不变，并加固非 dict payload；
- 新增 8 个 hook 测试：协议位置与透传、三类警告与按会话去重、marker 有界裁剪、barrier 强制 read-read-write-write 交错的并发回归、6 进程并行首次告警；
- troubleshooting 文档新增「Schema 压缩没有统计记录」小节，按宿主准确描述 Schema 压缩接入方式：cosh/Cosh-NG 经 BeforeModel；OpenCode 经 `tool.definition` 插件 Hook（MCP 工具不经过）；Qwen Code 的清单声明了 BeforeModel 条目，但当前版本宿主注册器跳过未知事件名，Hook 不运行、零记录为预期；user-manual 任务索引与能力表、framework-integration 支持矩阵与图例同步与同一事实对齐；均含中文镜像。

**B. Gemini 包装形态增量测试与文档（核心已在 main）**

- 3 个补充单元测试：多声明包装保持名称/顺序并移除声明级 title；非数组 `functionDeclarations` 原样放行；零收益包装返回原值；
- `compress-schema --batch` Gemini 包装形态 CLI 集成测试；
- CLI 参考文档（中/英）输入形态说明（含 `parametersJsonSchema` 变体）。

**宿主侧写回（已由当前 base 解决）**：此前 copilot-shell 的 `fromHookLLMRequest` 不会把 hook 返回的 `config.tools` 写回实际请求。当前 base 已包含 `afd3ff58`（fix(cosh): merge hook-modified tools into request）：override 白名单加入 `tools`，BeforeModel hook 改写后的工具集随其余 override 字段一起写回请求。reviewer 已实测确认压缩结果在 copilot-shell 上正常生效（stats 记录亦正常），无需再在宿主侧另行修复。

## 测试情况

**实际执行的命令与结果（当前 HEAD）**：

- `cargo test -p tokenless-schema`：26 passed / 0 failed（含本 PR 新增 Gemini 包装单元测试）
- `cargo test -p tokenless-cli --test cli_integration`：39 passed / 0 failed（含本 PR 新增 Gemini 包装 CLI 集成测试）
- Python hook 测试：`test_compress_schema_hook` 17/17（Python 3.9 与 3.12 双版本）、`test_compress_response_hook` 24/24、`test_resolve_agent_id` 7/7；`py_compile` 对 Python 3.8 / 3.9 / 3.12 通过
- 文档门禁：`scripts/docs-lint.sh`（命名规范 / en-zh 目录一致性）、`scripts/docs-link-check.py` 通过

**复现问题 → 验证修复（真实 hook + 隔离 HOME）**：

- 无工具声明负载：修复前完全静默；修复后首次给出 stderr + systemMessage 警告，同会话去重，新会话重新告警；合法空 tools 数组保持静默；
- 并发竞态：barrier 强制 read-read-write-write 交错下，修复前 20/20 次丢失一个 session key（丢失的会话下一轮重复告警），修复后 0/20；临时禁用锁时新增回归测试按预期失败（回归捕捉有效）；
- 6 个 hook 进程并行首次告警：全部 session key 保留，第二轮全部静默。

**环境概要**：Linux x86_64；Rust/cargo 1.94.1；Python 3.8.17 / 3.9.21 / 3.12.9。

**未运行项及原因**：

- 真实 cosh / Cosh-NG / Qwen Code / OpenCode 会话联调：本环境无法启动真实 LLM 会话；宿主事件发射与载荷形态通过阅读宿主源码并以逐字段一致的 payload 等价验证；
- `test_rewrite_hook`（11 项）等既有失败为本机已安装真实 rtk 二进制干扰导致的既有环境性失败，干净基线上同样失败，与本 PR 无关。

---

> ## Rebase 记录（2026-08-19）
>
> 分支已 rebase 到最新 `main`（`2c95e53b`），`schema_compressor.rs` 的合并冲突已解决，force-push 完成。
>
> **冲突根因**：main 已合入同功能的并行实现 `625f0779`（feat(tokenless): compress Gemini tool schemas，含 `parametersJsonSchema` 支持）。rebase 时压缩器核心改动以 main 版本为准，原提交改写为 `test(tokenless): cover Gemini wrapper edge cases and document input shapes`，仅保留 main 未覆盖的增量：
>
> - 3 个补充单元测试：多声明包装保持名称/顺序并移除声明级 title、非数组 `functionDeclarations` 原样放行、零收益包装返回原值；
> - `compress-schema --batch` 的 Gemini 包装形态 CLI 集成测试；
> - CLI 参考文档（中/英）输入形态说明（含 `parametersJsonSchema` 变体）。
>
> **rebase 后重新执行的验证（全部通过）**：`cargo fmt --check`、`cargo clippy --workspace --all-targets -- -D warnings`、`cargo test --workspace`（585 passed / 0 failed / 2 ignored，含上述新增测试）、Python hook 测试（test_compress_schema_hook 15 + test_compress_response_hook 24 + test_resolve_agent_id 7，共 46 项通过）、hook 文件 py_compile（Python 3.8 / 3.12）、`scripts/docs-lint.sh` 与 `scripts/docs-link-check.py`。CI 在新 HEAD 上全绿（Test tokenless 通过，其余组件检查因未改动按变更检测跳过）。
>
> 以下为本 PR 原始说明（描述 rebase 前的完整改动背景）。

### CODE REVIEW

File: src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py [hunk 1]
```diff
@@ -68,21 +152,42 @@ def main() -> None:
     except (json.JSONDecodeError, EOFError, ValueError):
         warn("failed to read BeforeModel payload. Passing through unchanged.")
         skip()
+    if not isinstance(input_data, dict):
```
[the remark this answers is not in the document]
Comment (Forrest-ly) [reply src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py:231]: 已采纳：已在 troubleshooting 文档（en/zh）「确认 BeforeModel Hook 已触发」小节补充该分支——现在小节列出全部三条一次性警告，第一条即"payload is not a JSON object"，并说明其含义是 Hook 收到的负载不是 JSON 对象。见 b50dd3e。

File: src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py [hunk 2]
```diff
@@ -68,21 +152,42 @@ def main() -> None:
...
+    # 3. Extract tools array.
+    tools, tools_declared = _extract_tools(input_data)
     if not tools:
+        # A host that declares no tools for this turn (empty canonical array)
+        # skips silently — that is a normal request. But a BeforeModel event
+        # with no tools field at any known position means there is nothing
+        # for schema compression to work on. That skip used to be silent, so
+        # "0 schema records" could not be told apart from "hook never ran";
+        # warn once per session instead.
+        if not tools_declared and _should_warn_no_tools(
```
[the remark this answers is not in the document]
Comment (Forrest-ly) [reply src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py:250]: 已采纳：与上一条合并在同一次文档更新中——troubleshooting 文档（en/zh）「确认 BeforeModel Hook 已触发」小节现在列出全部三条一次性警告，第二条即"payload carries no llm_request object"，并说明其含义是负载缺少 llm_request 对象。见 b50dd3e。

File: docs/user-guide/en/token-saving/tokenless/troubleshooting.md [hunk 3]
```diff
@@ -120,6 +120,36 @@ Confirm that `TOKENLESS_STATS_ENABLED=0` is not set unexpectedly and that any
...
 
+## Schema compression produces no statistics
+
+Schema compression runs on the BeforeModel hook before every model call; cosh, Cosh-NG, Qwen Code, and OpenCode register the hook. When there are no `compress-schema` records, check the following in order:
+
+### 1. Confirm there is something to compress
+
+Statistics only record invocations that save tokens; a result that is not smaller than the original is not recorded. Built-in tool descriptions are usually short (below the 256-character function and 160-character parameter truncation thresholds, with no `title` or `examples` to remove), so compression yields no savings and zero records are expected. Verify directly with the current tool declarations:
+
+```bash
```
[the remark this answers is not in the document]
Comment (Forrest-ly) [reply docs/user-guide/en/token-saving/tokenless/troubleshooting.md:137]: 已采纳：已把占位符替换为可直接运行的完整 JSON 数组示例（示例描述刻意超过 256 字符截断阈值，原样复制执行即可看到压缩效果），并明确注明"请替换为真实工具声明，不要保留占位文本、尖括号或外侧引号"（en/zh 同步修改）。未改为双引号写法：JSON 内部本身含双引号，在 POSIX shell 中用单引号包裹才是合法的，踩坑根源在于占位符被原样复制而非引号类型。见 b50dd3e。

File: src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py [hunk 4]
```diff
@@ -45,6 +55,80 @@ def _is_json_array(data: str) -> bool:
...
+    Best-effort by design: the marker lives under ``~/.tokenless`` and is
+    read/written with the same hardened helpers as other hook state, but any
+    failure falls back to warning — observability beats silence, and the
+    warning itself never affects the pass-through behaviour.
+    """
+    key = _session_warn_key(session_id)
+    try:
+        with open(_NO_TOOLS_WARN_MARKER, "r", encoding="utf-8") as handle:
+            if handle.read().strip() == key:
+                return False
```
[the remark this answers is not in the document]
Comment (Forrest-ly) [reply src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py]: Adopted. The marker now stores every warned session key — one key per line, most recent last, bounded to the newest 64 entries — instead of a single value, so a later session can no longer overwrite an earlier session's warning state; evicting the oldest entry only re-arms the warning for a long-idle session. The file keeps the same path and hardened-write semantics (0600, symlink-safe), and the read tolerates legacy single-key markers. Added two regression tests: interleaved A/B turns stay deduped after both sessions warn, and a pre-seeded over-bound marker is trimmed oldest-first while active keys survive. See b50dd3e.

File: docs/user-guide/en/token-saving/tokenless/troubleshooting.md [hunk 5]
```diff
@@ -120,6 +120,38 @@ Confirm that `TOKENLESS_STATS_ENABLED=0` is not set unexpectedly and that any
 custom database path remains under the real user home or selected data
 directory.
 
+## Schema compression produces no statistics
+
+Schema compression runs on the BeforeModel hook before every model call; cosh, Cosh-NG, Qwen Code, and OpenCode register the hook. When there are no `compress-schema` records, check the following in order:
```
Comment (ikunkun-sys) [inline docs/user-guide/en/token-saving/tokenless/troubleshooting.md]: [P1] 这里把 Qwen Code 和 OpenCode 描述成会注册 BeforeModel，但上游协议并非如此。Qwen Code v0.21.2/v0.21.14 的官方 HookEventName 没有 BeforeModel，注册器会跳过未知事件；实际加载此扩展时，清单中的 4 组定义也只注册了 3 组，所以 schema hook 和本 PR 新增的告警都不会运行。OpenCode v1.14.48 使用 tool.definition，且 MCP 工具路径未触发该 hook，也不走这套 BeforeModel 告警。因此这段说明会把 Qwen 的零记录误报为已支持且可诊断。官方契约：[Qwen hooks](https://qwenlm.github.io/qwen-code-docs/en/users/features/hooks/)、[Qwen HookEventName](https://github.com/QwenLM/qwen-code/blob/v0.21.14/packages/core/src/hooks/types.ts#L24-L69)、[Qwen registry](https://github.com/QwenLM/qwen-code/blob/v0.21.2/packages/core/src/hooks/hookRegistry.ts#L283-L299)、[OpenCode plugin API](https://github.com/anomalyco/opencode/blob/v1.14.48/packages/plugin/src/index.ts#L329-L332)。

File: src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py [hunk 6]
```diff
@@ -45,6 +64,97 @@ def _is_json_array(data: str) -> bool:
...
+    Every warned session key stays in the marker (bounded to the newest
+    ``_MAX_WARNED_SESSIONS`` entries), so concurrent sessions under one HOME
+    do not invalidate each other's dedup state. Best-effort by design: the
+    marker lives under ``~/.tokenless`` and is read/written with the same
+    hardened helpers as other hook state, but any failure falls back to
+    warning — observability beats silence, and the warning itself never
+    affects the pass-through behaviour.
+    """
+    key = _session_warn_key(session_id)
+    warned = _read_warned_sessions()
```
Comment (ikunkun-sys) [inline src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py]: [P2] 这里仍是无锁的 read-modify-truncate-write：两个 hook 进程可以同时读到相同旧集合，再分别写回并互相覆盖。用 barrier 强制该交错后，20/20 次都只保留了两个 session key 中的一个，丢失的会话下一轮会再次告警，违背 once-per-session 保证。现有 `test_concurrent_sessions_do_not_invalidate_each_other` 只是顺序执行 A/B/A/B，没有覆盖这个真实并发交错。

**Review by ikunkun-sys — CHANGES_REQUESTED** [review]
当前 HEAD 上有两项仍可复现的阻塞问题：Qwen Code/OpenCode 并不按文档所述注册 BeforeModel，以及 session marker 在真实并发下会丢失已记录的 key。另请同步更新 PR 描述：当前 base 已包含 afd3ff58，实测 cosh 的 translator → hook → translator 路径已能把压缩后的 config.tools 写回。

File: docs/user-guide/en/token-saving/tokenless/troubleshooting.md (hunk 5 above)
Comment (Forrest-ly) [reply docs/user-guide/en/token-saving/tokenless/troubleshooting.md]: 已采纳。已核对上游源码，两点均属实：

1. **Qwen Code**：v0.21.14 `packages/core/src/hooks/types.ts` 的 `HookEventName` 枚举没有 `BeforeModel`；v0.21.2 `hookRegistry.ts` 的 `processHooksConfiguration` 对未知事件名发出 warning 后直接 skip。因此扩展清单中的 `BeforeModel` 组在加载时被丢弃（4 组只注册 3 组），schema hook 与本 PR 新增告警都不会运行，Qwen 上零记录属于预期。
2. **OpenCode**：schema 压缩走 `tool.definition` 插件 hook（逐工具定义），不走 BeforeModel；MCP 工具不经过该路径。

修复（commit abca69902，troubleshooting en/zh）：

- 「Schema 压缩没有统计记录 / produces no statistics」小节改为按宿主分条说明：cosh/Cosh-NG 通过 `BeforeModel` 运行（本节告警仅适用于它们）；OpenCode 通过 `tool.definition` 逐个压缩（MCP 工具无记录、BeforeModel 告警不适用）；Qwen Code 的 `BeforeModel` 条目会被当前版本注册器跳过，schema hook 不运行、零记录无法用本节诊断。
- 第 2 步「确认 BeforeModel Hook 已触发」的告警说明同步限定为 cosh / Cosh-NG。
- user-manual 与 framework-integration（en/zh）中的同类表述（"只有 cosh 和 Qwen Code 注册 Schema Hook" 等）一并更正，避免文档前后矛盾。

`scripts/docs-lint.sh` 与 `scripts/docs-link-check.py` 均通过。

File: src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py (hunk 6 above)
Comment (Forrest-ly) [reply src/tokenless/adapters/tokenless/common/hooks/compress_schema_hook.py]: 已采纳。先按 review 的方法复现确认：barrier 强制 read-read-write-write 交错下，修复前 20/20 次丢失两个 session key 中的一个。

修复（commit df112982d）：把 `_should_warn_no_tools` 的 check-and-record 整体放入对专用 sidecar 锁文件的独占 `flock` 中串行化：

- 锁放在与 marker 分离的 sidecar 文件上，marker 写入不会替换被锁的 inode；
- 进程退出或崩溃时内核自动释放 `flock`，不会留下死锁阻塞后续会话；
- 无 `fcntl` 的平台或锁文件创建失败时退回原有无锁 best-effort 语义，告警行为不受影响。

新增回归测试（原有顺序 A/B/A/B 测试保留）：

1. `test_racing_first_warnings_are_serialized_by_marker_lock`：在 read 中注入 barrier 强制该交错。修复后第二个 worker 被锁挡住、无法到达自己的 read，barrier 超时断开，两个 key 都保留且各自下一轮静默。已验证：临时禁用锁时该测试如预期失败（只剩一个 key）。
2. `test_parallel_hook_processes_keep_every_session_key`：6 个 hook 进程并行首次告警，断言 marker 保留全部 key、第二轮全部静默。

测试：Python 3.9.21 与 3.12.9 下 `test_compress_schema_hook` 17/17 通过；`test_compress_response_hook` + `test_resolve_agent_id` 31/31 通过；`py_compile` 通过。

File: docs/user-guide/en/token-saving/tokenless/framework-integration.md [hunk 7]
```diff
@@ -22,6 +22,8 @@ Python framework package that application developers install and register explic
 
 “—” means that the current adapter does not register that capability. The corresponding Tokenless CLI command may still be available.
 
+Schema compression reaches the model path differently per host: cosh and Cosh-NG fire the `BeforeModel` hook; OpenCode compresses each tool definition through its `tool.definition` plugin hook (MCP tools do not pass through that hook); Qwen Code's manifest declares a `BeforeModel` hook, but current Qwen Code releases skip that unknown event name at registration, so the schema hook does not run there.
```
Comment (ikunkun-sys) [inline docs/user-guide/en/token-saving/tokenless/framework-integration.md]: [P2] 这里已正确说明 Qwen Code 会跳过未知的 `BeforeModel`、Schema Hook 实际不运行，但同页上方支持矩阵仍在 Qwen Code 行把 Schema 标为 ✅。该矩阵用于表达宿主能力，且紧随其后的说明只把 `—` 解释为未注册能力，因此两处会给用户相反结论。Qwen Code 最新 v0.21.14 的 `HookEventName` 仍没有 `BeforeModel`，所以当前矩阵仍会把不可用能力标成支持。

**Review by ikunkun-sys — CHANGES_REQUESTED** [review]
上一轮两项 inline blocker 已验证修复：marker 的 `flock` 覆盖了真实并发交错，宿主协议正文也已与 Qwen Code v0.21.14 / OpenCode v1.18.19 对齐。当前 HEAD 仍有三项需要处理：

1. 支持矩阵仍把 Qwen Code 的 Schema 标为 ✅，与本次新增正文和实际上游能力相互矛盾（见 inline）。
2. `df112982` 修复本 PR `7f19a5bd` 引入的并发缺陷，`abca6990` 也包含对本 PR 新增错误文档的修正；它们作为独立提交违反 `AGENTS.md` §13 的同 PR fixup/autosquash 规则。
3. 上一轮 review body 指出的 PR 元数据仍未更新：标题和正文仍把已由 #2663 合入的 Gemini 压缩核心列为本 PR 改动，并保留 cosh 不会回写 `config.tools` 的过期遗留项。

当前 HEAD 的针对性 Python 测试 17/17 通过；禁用锁后确定性竞态测试按预期失败。

File: docs/user-guide/en/token-saving/tokenless/framework-integration.md (hunk 7 above)
Comment (Forrest-ly) [reply docs/user-guide/en/token-saving/tokenless/framework-integration.md]: 已采纳。矩阵与正文确实给出了相反结论，已统一（commit 306d67f74，framework-integration en/zh）：

- 支持矩阵 Qwen Code 行 Schema 列 ✅ → `—`：矩阵表达宿主实际可用能力，当前 Qwen Code 版本（含 v0.21.14，`HookEventName` 仍无 `BeforeModel`）不会运行 Schema Hook，不应标为支持；
- `—` 图例扩展为"能力不可用：当前 Adapter 没有注册，或当前宿主版本不会运行"（原图例只覆盖前一种情况，正是本次矛盾的根源）；
- 正文 Schema 说明段补充"因此矩阵标记为不可用"，并注明清单仍保留该注册条目——未来 Qwen Code 版本实现 `BeforeModel` 事件后会自动生效，无需再改 Adapter。

`scripts/docs-lint.sh` 与 `scripts/docs-link-check.py` 均通过。

**Review by ikunkun-sys — CHANGES_REQUESTED** [review]
当前 HEAD `5055d1a` 上，上一轮代码与文档问题均已验证修复：marker 的 `flock` 覆盖真实并发交错；Qwen Code/OpenCode 能力说明及支持矩阵与最新上游契约一致；PR 标题和正文也已更新。针对性 Python hook 测试 17/17、Gemini schema 测试 7/7、CLI 集成测试、双语文档 lint/link 均通过。

仍有一项需要处理：

[P2] `5055d1a` 修复了本 PR 首个提交 `48c2302` 引入的 marker 并发问题，提交说明也明确写着 “Address PR review feedback”。根据 `AGENTS.md` §13，同一 PR 内引入的缺陷必须通过 fixup/autosquash 折叠到原提交，不能保留独立 fix commit。请将 `5055d1a` 折叠进 `48c2302`。

**Review by ikunkun-sys — APPROVED** [review]
重新 review 当前 HEAD `ba06d44`：未发现剩余阻塞问题。

本次 force-push 仅重写提交历史，新旧 HEAD 的文件 tree 完全一致；并发 marker 修复已按 `AGENTS.md` §13 autosquash 进首个功能提交，上一轮唯一 finding 已关闭。Qwen Code v0.21.15 与 OpenCode v1.18.21 的最新上游契约仍与当前文档一致。

验证：`cargo clippy --workspace --all-targets --locked -- -D warnings` 通过；`cargo test --workspace --locked` 601 passed / 2 ignored；Python schema hook 17/17；双语文档 lint、链接检查与 `git diff --check` 均通过。
